src/app/clusters/scenes/SceneTableImpl.cpp

```diff
@@ -0,0 +1,680 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/clusters/scenes/SceneTableImpl.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <stdlib.h>
+
+namespace chip {
+namespace scenes {
+
+enum SceneImplTLVTag
+{
+    kTagSceneCount = 1,
+    kTagNext,
+};
+
+using SceneTableEntry = DefaultSceneTableImpl::SceneTableEntry;
+using SceneStorageId  = DefaultSceneTableImpl::SceneStorageId;
+using SceneData       = DefaultSceneTableImpl::SceneData;
+
+struct FabricHavingSceneList : public CommonPersistentData::FabricList
+{
+    // This implementation uses the same key as the GroupFabricList from GroupDataProviderImpl to avoid duplicating the list in
+    // memory. If a different GroupDataProvider implementation is used, it will create the list in flash memory.
+    CHIP_ERROR UpdateKey(StorageKeyName & key) override
+    {
+        key = DefaultStorageKeyAllocator::GroupFabricList();
+        return CHIP_NO_ERROR;
+    }
+};
+
+// Worst case tested: Add Scene Command with EFS using the default SerializeAdd Method. This yielded a serialized scene of 212bytes
+// when using the OnOff, Level Control and Color Control as well as the maximal name length of 16 bytes. Putting 256 gives some
+// slack in case different clusters are used. Value obtained by using writer.GetLengthWritten at the end of the SceneTableData
+// Serialize method.
+static constexpr size_t kPersistentSceneBufferMax = 256;
+
+struct SceneTableData : public SceneTableEntry, PersistentData<kPersistentSceneBufferMax>
+{
+    FabricIndex fabric_index = kUndefinedFabricIndex;
+    SceneIndex index         = 0;
+    bool first               = true;
+
+    SceneTableData() : SceneTableEntry() {}
+    SceneTableData(FabricIndex fabric) : fabric_index(fabric) {}
+    SceneTableData(FabricIndex fabric, SceneIndex idx) : fabric_index(fabric), index(idx) {}
+    SceneTableData(FabricIndex fabric, SceneStorageId storageId) : SceneTableEntry(storageId), fabric_index(fabric) {}
+    SceneTableData(FabricIndex fabric, SceneStorageId storageId, SceneData data) :
+        SceneTableEntry(storageId, data), fabric_index(fabric)
+    {}
+
+    CHIP_ERROR UpdateKey(StorageKeyName & key) override
+    {
+        VerifyOrReturnError(kUndefinedFabricIndex != fabric_index, CHIP_ERROR_INVALID_FABRIC_INDEX);
+        key = DefaultStorageKeyAllocator::FabricSceneKey(fabric_index, index);
+        return CHIP_NO_ERROR;
+    }
+
+    void Clear() override { mStorageData.Clear(); }
+
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const override
+    {
+        CharSpan NameSpan(mStorageData.mName, mStorageData.mNameLength);
+        TLV::TLVType container;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, container));
+
+        // Scene ID
+        ReturnErrorOnFailure(writer.Put(TLV::ContextTag(kTagSceneEndpointID), static_cast<uint16_t>(mStorageId.mEndpointId)));
```

Comment:
Why is this static_cast needed?  It shouldn't be needed.  We chould be writing the actual data we have, whatever that is.

Reply:
Not needed indeed, removed in latest commit.